generator/booking_dataset.cpp

```diff
@@ -244,22 +277,67 @@ double BookingDataset::ScoreByLinearNormDistance(double distance)
   return 1.0 - distance / kDistanceLimitInMeters;
 }
 
-void BookingDataset::LoadHotels(string const & path)
+void BookingDataset::LoadHotels(istream & src, string const & addressReferencePath)
 {
   m_hotels.clear();
 
-  if (path.empty())
-    return;
+  for (string line; getline(src, line);)
+    m_hotels.emplace_back(line);
 
-  ifstream src(path);
-  if (!src.is_open())
+  if (!addressReferencePath.empty())
   {
-    LOG(LERROR, ("Error while opening", path, ":", strerror(errno)));
-    return;
+    LOG(LINFO, ("Match addresses for booking objects", addressReferencePath));
+    Platform & platform = GetPlatform();
+    string const backupPath = platform.WritableDir();
+    platform.SetWritableDirForTests(addressReferencePath);
+
+    vector<platform::LocalCountryFile> localFiles;
+    platform::FindAllLocalMapsInDirectoryAndCleanup(platform.WritableDir(), 0 /* version */,
+                                                    -1 /* latestVersion */, localFiles);
+
+    Index index;
+
+    for (platform::LocalCountryFile const & localFile : localFiles)
+    {
+      LOG(LINFO, ("Found mwm:", localFile));
+      try
+      {
+        index.RegisterMap(localFile);
+      }
+      catch (RootException const & ex)
+      {
+        CHECK(false, ("Bad mwm file:", localFile));
+      }
+    }
+
+    search::ReverseGeocoder const coder(index);
+
+    size_t matchedNum = 0;
+    size_t emptyAddr = 0;
+    for (Hotel & hotel : m_hotels)
+    {
+      search::ReverseGeocoder::Address addr;
+      coder.GetNearbyAddress(MercatorBounds::FromLatLon(hotel.lat, hotel.lon), addr);
+      hotel.street = addr.GetStreetName();
+      hotel.houseNumber = addr.GetHouseNumber();
+
+      if (hotel.address.empty())
+        ++emptyAddr;
+      if (!(hotel.street + hotel.houseNumber).empty())
+        ++matchedNum;
+    }
+    LOG(LINFO,
+        ("Num of hotels:", m_hotels.size(), "matched:", matchedNum, "Empty addresses:", emptyAddr));
+    platform.SetWritableDirForTests(backupPath);
   }
 
-  for (string line; getline(src, line);)
-    m_hotels.emplace_back(line);
+  size_t counter = 0;
+  for (auto const & hotel : m_hotels)
+  {
+    TBox b(TPoint(hotel.lat, hotel.lon), TPoint(hotel.lat, hotel.lon));
+    m_rtree.insert(make_pair(b, counter));
```

Comment:
Do you need to clean BookingDataset state at the beginning of the method? I mean, m_rtree should be cleaned, right?